README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# coefa

The goal of `coefa` is to provide a calculation program for the method
of Meta Analysis of Factor Analysis Based on Co-occurrence Matrices. The
`coefa` package is an effective tools which can be used to solve the
factor structure (i.e. inner structure of a construct, or scale) debate
in several disciplines, such as psychology,
psychiatry,management,education et al.

## Installation

You can install the development version of `coefa` like so:

``` r
install.packages("coefa")
```

## Usage

This is a basic example which shows you how to solve a inner structure
debate:

### Read

First, the factor loading matrices data from a line of primary studies
should be imported into the R environment.These data(factor loading
matrices) should be stored using the type of list in R.This work can be
executed manually or using using the `coefa_read()`function of the
`coefa` package.

``` r
library(coefa)
#> 载入需要的程辑包：openxlsx
#> 载入需要的程辑包：psych
```

The function `coefa_read()` provides an effective way to read data files
from a folder in a computer. The user can read several types of files by
specifying the parameters of the function. What should be noted is that
the miss values in data files will be replaced with the number of 0.

``` r
#Supposing that the type of data are xlsx files.
matrices.withoutNa<-coefa_read(type = "xlsx")
```

\*\*NOTE: Although the`coefa_read()` function can help you quickly read
the data in the folder, there are two points should be noted: (1) The
path of the stored file should be consistent with your workspace. (2)
The file in the folder should have the same file formats, and different
files should be set with different parameters.

### Trim factor loading matrices

In the step, all the factor loading matrices will be trimmed using the
Shafer’s(2005) method or the Loeber and Schmaling’s method (1985). And
the cutoff values(e.g., 0.3, 0.4, 0.5) can be given here according to
the users’ consideration.

``` r
#Suppose matrices.withoutNa is obtained by coefa_read function
mx1<-matrix(c(0.1,0.2,0.3,0.4,0.5,0.6),nrow = 3,byrow = TRUE)
mx2<-matrix(c(0.6,0.5,0.4,0.3,0.2,0.1),nrow = 3,byrow = TRUE)
matrices.withoutNa<-list(mx1,mx2)
#Take the Loeber&Schmaling(1985) method, the cutoff value is 0.4 as an example.The result is that values in the matrix greater than or equal to 0.4 will become 1, and less than 0.4 will become 0.
matrices.tflm<-coefa_tflm(matrices.withoutNa,methodE = "ls",cutoff = 0.4)
```

### Generate co-occurrence matrices

In this step, the function `coefa_gcm()` will be used to generate the
co-occurrence matrix.

``` r
matrices.gcm<-coefa_gcm(matrices.tflm)
```

### Aggregate co-occurrence matrices

In this step, aggregated co-occurrence matrix will be obtained using the
`coefa_acm()` in which the aggregation algorithm will be executed. Here,
you can set the parameter `samplesize = TURE` to add the weights to all
studies. The sample sizes will be valued by the `sz1` variable If
`samplesize = FALSE`, no weight will be considered in the aggregation
process. When this step finished, a final aggregated co-occurrence
matrix (weighted or unweighted by sample size) will be calculated.

``` r
#Assume that the sample sizes of the factor loading matrices for the two studies are 100 and 200, respectively.
sz1<-c(100,200)
matrices.acm<-coefa_acm(matrices.gcm,sz=sz1,samplesized=TRUE)
```

### Summary

The function `coefa_summary()` provides a preliminary screening and
suggestion for the later factor analysis. The results of Scree plot and
Kaiser’s criterion will be plotted by this function. Furthermore, this
function will test the aggregated co-occurrence matrix, and return that
whether it is a positive matrix not.

``` r
coefa_summary(matrices.acm,fa="pc")
```

### Factor analysis

Finally, the function `coefa_fa` will provide the choice for factor
extraction under the condition of co-occurrence matrix. The MDS, EFA,
PCA can be a choice, and the function will generate a plot for your
choice. A typical setting comes as follows.

``` r
coefa_fa(matrices.acm,nfactors = 6,methodcoefa = "EFA",rotate = "varimax",fm="pa")
```


NEWS.md
## **Changes in version 1.0.3**

-   Updated Introduction

## **Changes in version 1.0.2**

-   Added a function to fix the value of diagonal in similarity matrix

-   Added the function of `coefa_acm` ,the raw co-occurrence matrix can be displayed

-   Updated Introduction


R/coefa_tflm.R
#' Trimming factor loading matrices
#'
#' The original factor loading matrices after removing nulls will be trimmed through
#' this function.At the end all the matrices only have two elements,0 or 1.
#' @usage coefa_tflm(x,methodE=c("s","ls"),cutoff=c(0.3,0.4,0.5))
#' @param x A list of multiple original factor loading matrices included without
#' NA.
#' @param methodE The methods for trimming multiple factor loading matrices from
#'  the original study."s" and "ls" are two methods of trimming the original factor
#'  loading matrix, which are widely used in coefa studies.The s = Shafer’s(2005)
#'  method; ls = the Loeber and Schmaling’s method(1985).
#' @param cutoff The Criteria for trimming factor loadings matrices. When greater
#'  than the cutoff value, a factor loading is assigned as 1; otherwise, it
#'  assigned as 0. And the cutoff value can be given as 0.3, or 0.4, or 0.5.
#' @details This function is mainly used to discard the intensity information of
#' the original factor loading matrices and to retain the direction information
#' which make it into binary data.
#'
#' During the coefa trimming process,the strength information of the factor
#' loading matrices are given up,and the direction information is preserved.
#' "s" and "ls" are two different methods in the coefa study.
#'
#' If method="s", the factor loading matrices will be trimmed using the
#' Shafer's(2005)method, that is, only the highest salient loadings of each variable
#' in each factor are retained. If method="ls", the factor loading matrices will
#' be trimmed using the Loeber and Schmaling's (1985) method,all insignificant
#' factor loadings will be valued as 0.
#'
#' Cutoff is just some fixed criterion for trimming the factor loading matrix.
#' For example:if cutoff=0.3,elements in the factor loading matrix greater
#' than 0.3 will be given the value of 1,and elements less than 0.3 will be
#' replaced with 0.
#'
#' @return A list contained those trimmed factor loading matrices.
#' @export
#'
#' @references
#' Loeber,R., & Schmaling, K. B. (1985). Empirical evidence for overt and covert
#' patterns of antisocial conduct problems: a metaanalysis. Journal of abnormal
#'  child psychology, 13(2), 337--353.
#'
#' Shafer,A. B.(2005). Meta-analysis of the Brief Psychiatric Rating Scale
#' factor structure. Psychological Assessment, 17(3),324--335.
#'
#' Shafer,A. B. (2006). Meta-analysis of the factor structures of four depression
#'  questionnaires: Beck, CES-D, Hamilton, and Zung. Journal of clinical
#'  psychology, 62(1), 123--146.
#' @examples
#' mx1<-matrix(c(0.1,0.2,0.3,0.4),nrow=2)
#' mx2<-matrix(c(0.4,0.3,0.2,0.1),nrow=2)
#' list1<-list(mx1,mx2)
#' #Trim matrices using the Shafer's method ,cutoff is 0.3.
#' matrices.tflm<-coefa_tflm(list1,methodE="s",cutoff=0.3)
coefa_tflm<-function(x,methodE=c("s","ls"),cutoff=c(0.3,0.4,0.5)){
  #Loeber and Schmaling's method
  ##cutoff=0.3
  af1 <- function(x){
    for (i in 1:nrow(x)){
      for(j in 1:ncol(x)){
        if (x[i,j]>=0.3){
          x[i,j]=1
        }
        else{x[i,j]=0}
      }
    }
    return(x)
  }
  ##cutoff=0.4
  af2 <- function(x){
    for (i in 1:nrow(x)){
      for(j in 1:ncol(x)){
        if (x[i,j]>=0.4){
          x[i,j]=1
        }
        else{x[i,j]=0}
      }
    }
    return(x)
  }
  ##cutoff=0.5
  af3 <- function(x){
    for (i in 1:nrow(x)){
      for(j in 1:ncol(x)){
        if (x[i,j]>=0.5){
          x[i,j]=1
        }
        else{x[i,j]=0}
      }
    }
    return(x)
  }
  #Shafer's method
  ##cutoff=0.3
  AF1 <- function(x){
    for (i in 1:nrow(x)){
      ma=max(x[i,])
      for(j in 1:ncol(x)){
        if (x[i,j]==ma & x[i,j]>=0.3){
          x[i,j]=1
        }
        else{x[i,j]=0}
      }
    }
    return(x)
  }
  ##cutoff=0.4
  AF2 <- function(x){
    for (i in 1:nrow(x)){
      ma=max(x[i,])
      for(j in 1:ncol(x)){
        if (x[i,j]==ma & x[i,j]>=0.4){
          x[i,j]=1
        }
        else{x[i,j]=0}
      }
    }
    return(x)
  }
  ##cutoff=0.5
  AF3 <- function(x){
    for (i in 1:nrow(x)){
      ma=max(x[i,])
      for(j in 1:ncol(x)){
        if (x[i,j]==ma & x[i,j]>=0.5){
          x[i,j]=1
        }
        else{x[i,j]=0}
      }
    }
    return(x)
  }
  if(methodE=="s"){
    for (i in 1:length(x)) {
      if(cutoff=="0.3"){
        x[[i]]<-AF1(x[[i]])
      }else if(cutoff=="0.4"){
        x[[i]]<-AF2(x[[i]])
      }else if(cutoff=="0.5"){
        x[[i]]<-AF3(x[[i]])
      }
    }
    re2<-x
  }
  else if(methodE=="ls"){
    for (i in 1:length(x)) {
      if(cutoff=="0.3"){
        x[[i]]<-af1(x[[i]])
      }else if(cutoff=="0.4"){
        x[[i]]<-af2(x[[i]])
      }else if(cutoff=="0.5"){
        x[[i]]<-af3(x[[i]])
      }
    }
    re2<-x
  }
  return(re2)
}
